File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode3217DeleteNodesFromLinkedListPresentInArray.cpp
#include "LeetCode3217DeleteNodesFromLinkedListPresentInArray.h"

#include <unordered_map>
#include <unordered_set>

#include "LeetCode206ReverseLinkedList.h"
using namespace std;

ListNode* LeetCode3217DeleteNodesFromLinkedListPresentInArray::modifiedList(vector<int>& nums, ListNode* head)
{
    unordered_set<int> numMap{};
    for (const int num : nums)
    {
        numMap.insert(num);
    }
    ListNode* previous=nullptr;
    auto node = head;
    auto newHead=head;
    while (node != nullptr)
    {
        if (numMap.contains(node->val))
        {
            if (previous!=nullptr)
            {
                previous->next = node->next;
                node = previous->next;
                continue;
            }
            newHead = node->next;
            node=node->next;
            continue;
        }
        previous = node;
        node = node->next;
    }
    return newHead;
}

File: LeetCodeProblemsGo/LeetCode3217DeleteNodesFromLinkedListPresentInArray/LeetCode3217DeleteNodesFromLinkedListPresentInArray.go
package LeetCode3217DeleteNodesFromLinkedListPresentInArray

import "LeetCodeProblemsGo/LinkedList"

type ListNode = LinkedList.ListNode

func modifiedList(nums []int, head *ListNode) *ListNode {
	numMap := make(map[int]any)
	for _, num := range nums {
		numMap[num] = struct{}{}
	}
	newHead := head
	var prev *ListNode = nil
	current := head
	for current != nil {
		if _, exists := numMap[current.Val]; exists {
			if prev == nil {
				newHead = current.Next
				current = newHead
				continue
			}
			prev.Next = current.Next
			current = current.Next
			continue
		}
		prev = current
		current = current.Next
	}
	return newHead
}
